vm_insnhelper.c

```diff
@@ -6372,47 +6372,73 @@ vm_ic_track_const_chain(rb_control_frame_t *cfp, IC ic, const ID *segments)
     RB_VM_LOCK_LEAVE();
 }
 
-// For JIT inlining
-static inline bool
-vm_inlined_ic_hit_p(VALUE flags, VALUE value, const rb_cref_t *ic_cref, const VALUE *reg_ep)
+static bool
+vm_ic_hit_p(const struct iseq_inline_constant_cache *ic, const VALUE *reg_ep)
 {
-    if ((flags & IMEMO_CONST_CACHE_SHAREABLE) || rb_ractor_main_p()) {
-        VM_ASSERT(ractor_incidental_shareable_p(flags & IMEMO_CONST_CACHE_SHAREABLE, value));
+    VALUE value = ic->value;
 
-        return (ic_cref == NULL || // no need to check CREF
-                ic_cref == vm_get_cref(reg_ep));
+    if (RB_UNLIKELY(UNDEF_P(value))) {
+        return false;
     }
-    return false;
-}
 
-static bool
-vm_ic_hit_p(const struct iseq_inline_constant_cache_entry *ice, const VALUE *reg_ep)
-{
-    VM_ASSERT(IMEMO_TYPE_P(ice, imemo_constcache));
-    return vm_inlined_ic_hit_p(ice->flags, ice->value, ice->ic_cref, reg_ep);
+    VALUE flags = vm_icc_flags(ic);
+
+    if (!(flags & CONST_CACHE_SHAREABLE) || rb_ractor_main_p()) {
+        // There is a small window of time during which IC->value is set, but the
+        // CONST_CACHE_SHAREABLE flag hasn't yet been set.
+        // So before failing we MUST check `rb_ractor_shareable_p` again.
+        // See `vm_ic_update`.
+        if (!rb_ractor_shareable_p(value)) {
+            return false;
+        }
+    }
+    else {
+        VM_ASSERT(ractor_incidental_shareable_p(flags & CONST_CACHE_SHAREABLE, value));
+    }
+
+    const rb_cref_t *ic_cref = vm_icc_cref(ic);
+    return (ic_cref == NULL || // no need to check CREF
+            ic_cref == vm_get_cref(reg_ep));
 }
 
 // YJIT needs this function to never allocate and never raise
 bool
 rb_vm_ic_hit_p(IC ic, const VALUE *reg_ep)
 {
-    return ic->entry && vm_ic_hit_p(ic->entry, reg_ep);
+    return vm_ic_hit_p(ic, reg_ep);
 }
 
 static void
 vm_ic_update(const rb_iseq_t *iseq, IC ic, VALUE val, const VALUE *reg_ep, const VALUE *pc)
 {
     if (ruby_vm_const_missing_count > 0) {
         ruby_vm_const_missing_count = 0;
-        ic->entry = NULL;
         return;
     }
 
-    struct iseq_inline_constant_cache_entry *ice = IMEMO_NEW(struct iseq_inline_constant_cache_entry, imemo_constcache, 0);
-    RB_OBJ_WRITE(ice, &ice->value, val);
-    ice->ic_cref = vm_get_const_key_cref(reg_ep);
-    if (rb_ractor_shareable_p(val)) ice->flags |= IMEMO_CONST_CACHE_SHAREABLE;
-    RB_OBJ_WRITE(iseq, &ic->entry, ice);
+    VALUE cache_val = val;
+
+    const rb_cref_t *cref = vm_get_const_key_cref(reg_ep);
+    if (RB_UNLIKELY(cref)) {
+        struct iseq_inline_constant_cache_entry *ice = IMEMO_NEW(struct iseq_inline_constant_cache_entry, imemo_constcache, 0);
+        RB_OBJ_WRITE(ice, &ice->value, val);
+        ice->ic_cref = cref;
+        cache_val = (VALUE)ice;
+    }
```

Comment:
Should add a short comment to explain what this bit does. What does it mean for cref to be null or not null?